File: src/main/java/org/trostheide/generativeArt/GenerativeRibbon.java
package org.trostheide.generativeArt;

import org.trostheide.generativeArt.core.ArtGenerator;
import org.trostheide.generativeArt.core.ParameterDefinition;

import java.io.FileWriter;
import java.io.IOException;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class GenerativeRibbon implements ArtGenerator {

    // Default Configuration
    private static final int NUM_LINES = 6000;
    private static final double MAX_T = 25.0;
    private static final double SCALE = 2.0;

    // Instance variables for calculations
    private double currentScale = SCALE;
    private double centerX = 500;
    private double centerY = 500;

    private static final DecimalFormat df;
    static {
        df = new DecimalFormat("#.##");
        df.setDecimalFormatSymbols(DecimalFormatSymbols.getInstance(Locale.ENGLISH));
    }

    @Override
    public String getId() {
        return "generative-ribbon";
    }

    @Override
    public String getDisplayName() {
        return "Generative Ribbon";
    }

    public List<ParameterDefinition> getParameterDefinitions() {
        return List.of(
                ParameterDefinition.selection("Preset", "Custom", 
                    java.util.Arrays.asList("Custom", "Dense Coil", "Sparse Wire", "Long Thread"), 
                    "Select a predefined style"),
                ParameterDefinition.integer("Lines", NUM_LINES, 500, 20000, "Density of the ribbon"),
                ParameterDefinition.doubleVal("Length (Max T)", MAX_T, 5.0, 100.0, "Length of the ribbon"),
                ParameterDefinition.doubleVal("Scale", SCALE, 0.5, 5.0, "Zoom level"),
                ParameterDefinition.integer("Colors", 1, 1, 6, "Number of plotter layers"));
    }

    @Override
    public boolean onParameterChanged(String paramName, Object newValue, Map<String, Object> currentValues) {
        if ("Preset".equals(paramName) && newValue instanceof String) {
            String preset = (String) newValue;
            switch (preset) {
                case "Dense Coil":
                    currentValues.put("Lines", 12000);
                    currentValues.put("Length (Max T)", 50.0);
                    currentValues.put("Scale", 3.0);
                    return true;
                case "Sparse Wire":
                    currentValues.put("Lines", 2000);
                    currentValues.put("Length (Max T)", 15.0);
                    currentValues.put("Scale", 1.5);
                    return true;
                case "Long Thread":
                    currentValues.put("Lines", 8000);
                    currentValues.put("Length (Max T)", 80.0);
                    currentValues.put("Scale", 2.5);
                    return true;
                case "Custom":
                default:
                    return false;
            }
        }
        if (!"Preset".equals(paramName) && !"Custom".equals(currentValues.get("Preset"))) {
            currentValues.put("Preset", "Custom");
            return true;
        }
        return false;
    }

    @Override
    public String generate(Map<String, Object> params) {
        // Read Params
        int numLines = (int) params.getOrDefault("Lines", NUM_LINES);
        double maxT = (double) params.getOrDefault("Length (Max T)", MAX_T);
        this.currentScale = (double) params.getOrDefault("Scale", SCALE);
        int numColors = (int) params.getOrDefault("Colors", 1);

        // Dimensions
        double width = 1000;
        double height = 1000;
        if (params.containsKey("width"))
            width = ((Number) params.get("width")).doubleValue();
        if (params.containsKey("height"))
            height = ((Number) params.get("height")).doubleValue();

        this.centerX = width / 2;
        this.centerY = height / 2;

        return generateSVG(numLines, maxT, width, height, numColors);
    }

    private String generateSVG(int numLines, double maxT, double width, double height, int numColors) {
        double minX = Double.MAX_VALUE, maxX = -Double.MAX_VALUE;
        double minY = Double.MAX_VALUE, maxY = -Double.MAX_VALUE;

        StringBuilder[] pathBuilders = new StringBuilder[numColors];
        for (int i = 0; i < numColors; i++) pathBuilders[i] = new StringBuilder();

        for (int i = 0; i < numLines; i++) {
            double t = (double) i / numLines * maxT;

            Point3D p1_3d = calculatePathA(t);
            Point3D p2_3d = calculatePathB(t);

            Point2D p1 = project(p1_3d);
            Point2D p2 = project(p2_3d);

            if (p1.x() < minX) minX = p1.x();
            if (p1.x() > maxX) maxX = p1.x();
            if (p1.y() < minY) minY = p1.y();
            if (p1.y() > maxY) maxY = p1.y();

            if (p2.x() < minX) minX = p2.x();
            if (p2.x() > maxX) maxX = p2.x();
            if (p2.y() < minY) minY = p2.y();
            if (p2.y() > maxY) maxY = p2.y();

            int layerIndex = i % numColors;
            pathBuilders[layerIndex].append(String.format(java.util.Locale.US, "M %.2f %.2f L %.2f %.2f ", p1.x(), p1.y(), p2.x(), p2.y()));
        }

        java.util.List<String> rawPaths = new java.util.ArrayList<>();
        for (int i = 0; i < numColors; i++) {
            if (pathBuilders[i].length() > 0) {
                String layerColor = new String[] { "black", "#E31A1C", "#1F78B4", "#33A02C", "#FF7F00", "#6A3D9A" }[i % 6];
                rawPaths.add(String.format(java.util.Locale.US, "<path d='%s' stroke='%s' fill='none' stroke-width='1.0' vector-effect='non-scaling-stroke'/>", pathBuilders[i].toString(), layerColor));
            }
        }

        double margin = 50.0;
        double bboxWidth = maxX - minX;
        double bboxHeight = maxY - minY;

        if (bboxWidth <= 0) bboxWidth = 1;
        if (bboxHeight <= 0) bboxHeight = 1;

        double targetScale = Math.min((width - 2 * margin) / bboxWidth, (height - 2 * margin) / bboxHeight);
        double offsetX = (width - bboxWidth * targetScale) / 2.0 - minX * targetScale;
        double offsetY = (height - bboxHeight * targetScale) / 2.0 - minY * targetScale;

        StringBuilder out = new StringBuilder();
        out.append(String.format(java.util.Locale.US, "<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 %.1f %.1f' width='%.1f' height='%.1f'>\n", width, height, width, height));
        out.append(String.format(java.util.Locale.US, "<defs><clipPath id='pageClip'><rect width='%.1f' height='%.1f'/></clipPath></defs>\n", width, height));
        out.append(String.format(java.util.Locale.US, "<rect width='%.1f' height='%.1f' fill='white'/>\n", width, height));
        out.append("<g clip-path='url(#pageClip)'>\n");
        out.append(String.format(java.util.Locale.US, "  <g transform='translate(%.1f, %.1f) scale(%.4f)'>\n", offsetX, offsetY, targetScale));
        for (String p : rawPaths) {
            out.append("    ").append(p).append("\n");
        }
        out.append("  </g>\n");
        out.append("</g>\n</svg>");

        return out.toString();
    }

    // --- Math Helpers ---

    private Point3D calculatePathA(double t) {
        double baseX = Math.sin(t * 0.7) * 200;
        double baseY = Math.cos(t * 0.9) * 250;
        double baseZ = Math.sin(t * 0.5) * 200;

        double offsetX = Math.cos(t * 3.1 + 0.5) * 60;
        double offsetY = Math.sin(t * 3.3) * 60;
        double offsetZ = Math.cos(t * 3.7 + 1.0) * 60;

        return new Point3D(
                (baseX + offsetX) * currentScale,
                (baseY + offsetY) * currentScale,
                (baseZ + offsetZ) * currentScale);
    }

    private Point3D calculatePathB(double t) {
        double baseX = Math.sin(t * 0.7) * 200;
        double baseY = Math.cos(t * 0.9) * 250;
        double baseZ = Math.sin(t * 0.5) * 200;

        double offsetX = Math.cos(t * 3.1 + Math.PI) * 70;
        double offsetY = Math.sin(t * 3.3 + Math.PI) * 70;
        double offsetZ = Math.cos(t * 3.7 + Math.PI + 1.0) * 70;

        return new Point3D(
                (baseX + offsetX) * currentScale,
                (baseY + offsetY) * currentScale,
                (baseZ + offsetZ) * currentScale);
    }

    private Point2D project(Point3D p) {
        double perspective = 1000.0 / (1000.0 - p.z());
        double x2d = p.x() * perspective + centerX;
        double y2d = p.y() * perspective + centerY;
        return new Point2D(x2d, y2d);
    }

    // Records
    record Point3D(double x, double y, double z) {
    }

    record Point2D(double x, double y) {
    }

    public static void main(String[] args) {
        GenerativeRibbon generator = new GenerativeRibbon();
        String svg = generator.generate(Map.of());
        try (FileWriter fileWriter = new FileWriter("ribbon.svg")) {
            fileWriter.write(svg);
            System.out.println("Generated ribbon.svg");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: web/src/generators/GenerativeRibbon.js
import { Generator } from '../core/Generator';
import { ParameterDefinition } from '../core/ParameterDefinition';
import { SeededRandom } from '../utils/SeededRandom';

export class GenerativeRibbon extends Generator {
    constructor() {
        super();
        this.NUM_LINES = 6000;
        this.MAX_T = 25.0;
        this.SCALE = 2.0;
        this.currentScale = this.SCALE;
        this.centerX = 500;
        this.centerY = 500;
    }

    getId() {
        return "generative-ribbon";
    }

    getDisplayName() {
        return "Generative Ribbon";
    }

    getParameterDefinitions() {
        return [
            ParameterDefinition.selection("Preset", "Custom", ["Custom", "Dense Coil", "Sparse Wire", "Long Thread"], "Select a predefined style"),
            ParameterDefinition.integer("Lines", 6000, 500, 20000, "Density of the ribbon"),
            ParameterDefinition.doubleVal("Length (Max T)", 25.0, 5.0, 100.0, "Length of the ribbon"),
            ParameterDefinition.doubleVal("Scale", 2.0, 0.5, 5.0, "Zoom level"),
            ParameterDefinition.integer("Colors", 1, 1, 6, "Number of plotter layers")
        ];
    }

    onParameterChanged(paramName, newValue, currentValues) {
        if (paramName === "Preset" && typeof newValue === "string") {
            const preset = newValue;
            switch (preset) {
                case "Dense Coil":
                    currentValues["Lines"] = 12000;
                    currentValues["Length (Max T)"] = 50.0;
                    currentValues["Scale"] = 3.0;
                    return true;
                case "Sparse Wire":
                    currentValues["Lines"] = 2000;
                    currentValues["Length (Max T)"] = 15.0;
                    currentValues["Scale"] = 1.5;
                    return true;
                case "Long Thread":
                    currentValues["Lines"] = 8000;
                    currentValues["Length (Max T)"] = 80.0;
                    currentValues["Scale"] = 2.5;
                    return true;
                case "Custom":
                default:
                    return false;
            }
        }
        if (paramName !== "Preset" && currentValues["Preset"] !== "Custom") {
            currentValues["Preset"] = "Custom";
            return true;
        }
        return false;
    }

    generate(params) {
        const numLines = params["Lines"] || this.NUM_LINES;
        const maxT = params["Length (Max T)"] || this.MAX_T;
        // Default to scale from params or default
        this.currentScale = params["Scale"] !== undefined ? params["Scale"] : this.SCALE;
        const numColors = params["Colors"] || 1;

        const width = params["width"] || 1000;
        const height = params["height"] || 1000;

        this.centerX = width / 2;
        this.centerY = height / 2;

        return this.generateSVG(numLines, maxT, width, height, numColors);
    }

    generateSVG(numLines, maxT, width, height, numColors) {
        let minX = Infinity, maxX = -Infinity;
        let minY = Infinity, maxY = -Infinity;

        const pathBuilders = Array(numColors).fill("");

        for (let i = 0; i < numLines; i++) {
            const t = (i / numLines) * maxT;

            const p1_3d = this.calculatePathA(t);
            const p2_3d = this.calculatePathB(t);

            const p1 = this.project(p1_3d);
            const p2 = this.project(p2_3d);

            if (p1.x < minX) minX = p1.x;
            if (p1.x > maxX) maxX = p1.x;
            if (p1.y < minY) minY = p1.y;
            if (p1.y > maxY) maxY = p1.y;

            if (p2.x < minX) minX = p2.x;
            if (p2.x > maxX) maxX = p2.x;
            if (p2.y < minY) minY = p2.y;
            if (p2.y > maxY) maxY = p2.y;

            const layerIndex = i % numColors;
            pathBuilders[layerIndex] += `M ${p1.x.toFixed(2)} ${p1.y.toFixed(2)} L ${p2.x.toFixed(2)} ${p2.y.toFixed(2)} `;
        }

        const margin = 50.0;
        let bboxWidth = maxX - minX;
        let bboxHeight = maxY - minY;

        if (bboxWidth <= 0) bboxWidth = 1;
        if (bboxHeight <= 0) bboxHeight = 1;

        const targetScale = Math.min((width - 2 * margin) / bboxWidth, (height - 2 * margin) / bboxHeight);
        const offsetX = (width - bboxWidth * targetScale) / 2.0 - minX * targetScale;
        const offsetY = (height - bboxHeight * targetScale) / 2.0 - minY * targetScale;

        const layerColors = ["black", "#E31A1C", "#1F78B4", "#33A02C", "#FF7F00", "#6A3D9A"];

        let svg = `<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 ${width} ${height}' width='${width}' height='${height}'>\n`;
        svg += `<defs><clipPath id='pageClip'><rect width='${width}' height='${height}'/></clipPath></defs>\n`;
        svg += `<rect width='${width}' height='${height}' fill='white'/>\n`;
        svg += `<g clip-path='url(#pageClip)'>\n`;
        svg += `  <g transform='translate(${offsetX.toFixed(1)}, ${offsetY.toFixed(1)}) scale(${targetScale.toFixed(4)})'>\n`;
        
        for (let i = 0; i < numColors; i++) {
            if (pathBuilders[i].length > 0) {
                const color = layerColors[i % layerColors.length];
                svg += `    <path d='${pathBuilders[i]}' stroke='${color}' fill='none' stroke-width='1.0' vector-effect='non-scaling-stroke'/>\n`;
            }
        }
        
        svg += `  </g>\n`;
        svg += `</g>\n</svg>`;

        return svg;
    }

    calculatePathA(t) {
        const baseX = Math.sin(t * 0.7) * 200;
        const baseY = Math.cos(t * 0.9) * 250;
        const baseZ = Math.sin(t * 0.5) * 200;

        const offsetX = Math.cos(t * 3.1 + 0.5) * 60;
        const offsetY = Math.sin(t * 3.3) * 60;
        const offsetZ = Math.cos(t * 3.7 + 1.0) * 60;

        return {
            x: (baseX + offsetX) * this.currentScale,
            y: (baseY + offsetY) * this.currentScale,
            z: (baseZ + offsetZ) * this.currentScale
        };
    }

    calculatePathB(t) {
        const baseX = Math.sin(t * 0.7) * 200;
        const baseY = Math.cos(t * 0.9) * 250;
        const baseZ = Math.sin(t * 0.5) * 200;

        const offsetX = Math.cos(t * 3.1 + Math.PI) * 70;
        const offsetY = Math.sin(t * 3.3 + Math.PI) * 70;
        const offsetZ = Math.cos(t * 3.7 + Math.PI + 1.0) * 70;

        return {
            x: (baseX + offsetX) * this.currentScale,
            y: (baseY + offsetY) * this.currentScale,
            z: (baseZ + offsetZ) * this.currentScale
        };
    }

    project(p) {
        const perspective = 1000.0 / (1000.0 - p.z);
        const x2d = p.x * perspective + this.centerX;
        const y2d = p.y * perspective + this.centerY;
        return { x: x2d, y: y2d };
    }
}
